apps/src/sites/studio/pages/jotformLoader.js

```diff
@@ -0,0 +1,87 @@
+/**
+ * @file Entry point for a bundle to embed on pages where we are also letting JotForm embed
+ * an inline form using an iframe.  This bundle monitors the JotForm script and attempts
+ * to measure reachability of the JotForm domain, load success, load time, and reports metrics
+ * back to our system so we have a sense of how often this fails.
+ * We also attempt to improve the user experience when the form is unavailable.
+ */
+import logToCloud from '../../../logToCloud';
+
+function main(context) {
+  Promise.all([
+    checkJotFormFrameLoaded(context),
+    // Domains used by Jotform:
+    // cdn.jotfor.ms       https://cdn.jotfor.ms/images/calendar.png  817B
+    checkReachability('https://cdn.jotfor.ms/images/calendar.png'),
+    // www.jotform.com     https://www.jotform.com/favicon.ico        887B
+    checkReachability('https://www.jotform.com/favicon.ico')
+    // Not using these yet, they're likely to fall under the same policy as www.jotform.com
+    // events.jotform.com
+    // files.jotform.com
+    // form.jotform.com
+  ]).then(([jotFormFrameLoadedMs, cdnjotformsMs, wwwjotformcomMs]) => {
+    if (jotFormFrameLoadedMs === false) {
+      // Load failed if we specifcially got 'false'
+      logToCloud.addPageAction(logToCloud.PageAction.JotFormLoadFailed, {
+        route: `GET ${context.location.pathname}`,
+        reachedCdnjotforms: false !== cdnjotformsMs,
+        reachedWwwjotformcom: false !== wwwjotformcomMs,
+        cdnjotformsMs,
+        wwwjotformcomMs
+      });
+    } else {
+      logToCloud.addPageAction(logToCloud.PageAction.JotFormFrameLoaded, {
+        route: `GET ${context.location.pathname}`,
+        jotFormFrameLoadedMs,
+        cdnjotformsMs,
+        wwwjotformcomMs
+      });
+    }
+  });
+}
+
+/**
+ * Report page time when JotForm's embedded form successfully loads, or resolve(false) if it
+ * fails to load within five seconds.
+ * @param {Window} context
+ * @returns {Promise<number|false>} Always resolves, page time (ms) if load succeeded, `false` if not.
+ */
+function checkJotFormFrameLoaded(context) {
+  return new Promise(resolve => {
+    context.JotFormFrameLoaded = function() {
+      const time = performance.now();
+      if (timeoutKey) {
+        clearTimeout(timeoutKey);
+      }
+      console.log(`JotFormFrameLoaded fired at ${time}ms`);
+      resolve(time);
+    };
+
+    const timeoutKey = setTimeout(function() {
+      console.log(`JotForm failed to load in 5s`);
+      resolve(false);
+    }, 5000);
```

Comment:
I wonder if it might take more than 5 seconds on some slower computers and connections...

Reply:
Maybe. What would you recommend? We could turn this up, start gathering timings, and then tune it back down once we have a rough sense of our bell curve.

The risk of waiting too long to report is that the teacher gets impatient and closes/reloads the page before we've reported the failure to our servers.  I suppose I could attach an `onbeforeunload` event to try and catch that, but no guarantees it'll get the request off in time.

We could also set up a 'heartbeat' and record every second since page load that the form hasn't successfully loaded.  You'd get a sort of histogram, but I worry it gets complicated to separate actual failures from slow loads - maybe if we can capture sessions that don't include a load success event, that would do it?

Reply:
Agreed it's better to not complicate.

My gut says somewhere between 7 to 10 seconds, but I have absolutely no data for that :)  

Oh, I apparently chose 10 seconds for video reachability: https://github.com/code-dot-org/code-dot-org/pull/21885/files#diff-5180bfba62d8bd9766dc4a7ea63a943eR22

But the default for our reachability checks is 5 seconds: https://github.com/code-dot-org/code-dot-org/pull/21885/files#diff-96896ecf6bb59632fb9eba07868c0c08R3

In the case of this change, I guess we're not changing user functionality so much as gathering data on problems, and so choosing the lower timing - 5 seconds - will actually gather us more data about issues, so 5 seconds actually sounds good.